### PR TITLE
Fix CgroupsMemoryUsageObserver

### LINKED ISSUE #64652: CgroupsMemoryUsageObserver pessimizes current memory usage
```
2024.05.30 19:12:02.754186 [ 436184 ] {} <Trace> CgroupsMemoryUsageObserver: Read current memory usage 199.78 GiB from cgroups
```

```
➜  ~ date && free -mh && cat /sys/fs/cgroup/system.slice/clickhouse-server.service/memory.current && cat /sys/fs/cgroup/system.slice/clickhouse-server.service/memory.stat | head -n 2

Thu May 30 19:12:03 UTC 2024
               total        used        free      shared  buff/cache   available
Mem:           251Gi        25Gi        49Gi       3.0Mi       177Gi       224Gi
Swap:             0B          0B          0B

214512672768

anon 24153354240
file 188593889280
```

```
┌─formatReadableSize(plus(24153354240, 188593889280))─┐
│ 198.14 GiB                                          │
└─────────────────────────────────────────────────────┘
```

this leads to wrong OOM exceptions.

---

for cgroups v2 we use `memory.current` that includes page cache:

> memory.current - Shows the total amount of memory currently being used by the cgroup and its descendants. It includes page cache, in-kernel data structures such as inodes, and network buffers.
https://facebookmicrosites.github.io/cgroup2/docs/memory-controller.html

cc @rschu1ze

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improved memory accounting for cgroups v2 to exclude the amount occupied by the page cache.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

Fixes: #64652

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 56f7ab28f418a7143fb92e034e87b13bd2d7c7e4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65470/56f7ab28f418a7143fb92e034e87b13bd2d7c7e4/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Common/CgroupsMemoryUsageObserver.cpp [hunk 1]
```diff
@@ -22,90 +24,90 @@
...
+    uint64_t readMemoryUsage() override
+    {
+        std::lock_guard lock(mutex);
+        current_buf.rewind();
+        stat_buf.rewind();
+
+        uint64_t mem_usage = 0;
+        /// memory.current contains a single number
+        readIntText(mem_usage, current_buf);
+        mem_usage -= readMetricFromStatFile(stat_buf, "inactive_file");
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Common/CgroupsMemoryUsageObserver.cpp:102]: Not sure how much we can trust cgroups but perhaps put in a check that the result after subtraction is positive (or better return the original `mem_usage` if not).

File: src/Common/CgroupsMemoryUsageObserver.cpp [hunk 2]
```diff
@@ -132,29 +134,30 @@ std::optional<std::string> getCgroupsV2FileName()
...
         return {};
-    return {path};
+    return {default_cgroups_mount / "memory"};
 }
 
-std::pair<std::string, CgroupsMemoryUsageObserver::CgroupsVersion> getCgroupsFileName()
+std::pair<std::string, CgroupsMemoryUsageObserver::CgroupsVersion> getCgroupsPath()
 {
-    auto v2_file_name = getCgroupsV2FileName();
+    auto v2_file_name = getCgroupsV2Path();
```
Comment (rschu1ze) [inline src/Common/CgroupsMemoryUsageObserver.cpp:156]: Rename the variable to `v2_path`? (same in l. 160?)

**Review by rschu1ze — APPROVED** [review]

**Review by rschu1ze — APPROVED** [review]

**Comment by den-crane** [comment]
It's not backported to 24.3 ?

I have something similar

```
free -h
               total        used        free      shared  buff/cache   available
Mem:           251Gi        38Gi        29Gi       3.5Mi       186Gi       213Gi
Swap:             0B          0B          0B


Peak memory usage: 7.52 GiB.

Received exception from server (version 24.3.4):
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (total) exceeded: would use 226.44 GiB (attempt to allocate chunk of 0 bytes), maximum: 226.43 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.: While executing AggregatingTransform. (MEMORY_LIMIT_EXCEEDED)
```

Query which needs about 30GB RAM is unable to execute with `available 213Gi`

**Comment by nickitat** [comment]
> It's not backported to 24.3 ?
> 
> I have something similar
> 
> ```
> free -h
>                total        used        free      shared  buff/cache   available
> Mem:           251Gi        38Gi        29Gi       3.5Mi       186Gi       213Gi
> Swap:             0B          0B          0B
> 
> 
> Peak memory usage: 7.52 GiB.
> 
> Received exception from server (version 24.3.4):
> Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (total) exceeded: would use 226.44 GiB (attempt to allocate chunk of 0 bytes), maximum: 226.43 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.: While executing AggregatingTransform. (MEMORY_LIMIT_EXCEEDED)
> ```
> 
> Query which needs about 30GB RAM is unable to execute with `available 213Gi`

you could verify by grep-ing `'(CgroupsMemoryUsageObserver|AsynchronousMetrics|MemoryTrack)'` logs and also you should be using cgroups v2

**Comment by den-crane** [comment]
I think logging for `CgroupsMemoryUsageObserver` was implemented in 24.4

```
grep cgroup /proc/filesystems
nodev	cgroup
nodev	cgroup2
```

```
start
2024.09.13 15:52:06.842827 [ 202494 ] {0e042600-115f-4e6b-9505-9101ea87d829} <Debug> executeQuery: (from [::1]:41844) select count() FROM ( SELECT ad_key as `c


fail
2024.09.13 15:52:27.672780 [ 202494 ] {0e042600-115f-4e6b-9505-9101ea87d829} <Error> executeQuery: Code: 241. DB::Exception: Memory limit (total)
```

'(CgroupsMemoryUsageObserver|AsynchronousMetrics|MemoryTrack)'
for that period https://pastila.nl/?0021c336/540c081e5dc74bf6af4c3d3d4d5a373a#0tEYCArdnpecc//78jorzw==

**Comment by den-crane** [comment]
Though I cut log too early

https://pastila.nl/?00014c63/cbc2d89bcad5f94c9dda415bc3308b9c#gj7TpcjTjxBaisX32ppf2g==

2024.09.13 15:52:30.000790 [ 3551170 ] {} <Trace> AsynchronousMetrics: MemoryTracking: was 207.51 GiB, peak 226.43 GiB, free memory in arenas 14.30 GiB, will set to 43.74 GiB (RSS), difference: -163.77 GiB

**Comment by den-crane** [comment]
cgroups_memory_usage_observer_wait_time 
│ 15    │ 15      │       0 │ Polling interval in seconds to read the current memory usage from cgroups. Zero means disabled.

Probably I can use `cgroups_memory_usage_observer_wait_time=0` as a quick fix.

**Pull request merged**